### DISCUSSION TITLE
PorousFlow - mass balance wont close

### DISCUSSION THREAD

**Discussion by MatiasAllay** [idaholab/moose#18302] [Q&A Modules: General] [answered]
Hi, 
I'm sorry I have been asking so many simple questions. I appreciate a lot your help. I have a HM - RZ coordinates with a sink. The mass balance closes with hydro, but with mechanics somehow does not. Why is that? Here is my code

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 1
  ymax = 1 
  xmax = 100
  xmin = 0.2
[]

[Problem]
  coord_type = RZ
[]

[GlobalParams]
  displacements = 'disp_r disp_z'
  PorousFlowDictator = dictator
  biot_coefficient = 1.0
[]

[Variables]
  [disp_r]
  []
  [disp_z]
  []
  [porepressure]
  []
[]

[BCs]
  [sink]
    type = PorousFlowSink
    variable = porepressure
    flux_function = flux
    boundary = left
  []
  [r]
    type = FunctionDirichletBC
    variable = disp_r
    function = 0
    boundary = 'right left'
  []
  [z]
    type = FunctionDirichletBC
    variable = disp_z
    function = 0
    boundary = 'top bottom'
  []
[]

[Modules]
  [FluidProperties]
    [the_simple_fluid]
      type = SimpleFluidProperties
      bulk_modulus = 2E9
    []
  []
[]

[PorousFlowBasicTHM]
  coupling_type = HydroMechanical
  displacements = 'disp_r disp_z'
  porepressure = porepressure
  gravity = '0 0 0'
  fp = the_simple_fluid
  multiply_by_density = true
[]

[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    bulk_modulus = 1.0E7
    poissons_ratio = 0.25
  []
  [strain]
    type = ComputeAxisymmetricRZSmallStrain
  []
  [stress]
    type = ComputeLinearElasticStress
  []
  [porosity]
    type = PorousFlowPorosityConst
    porosity = 0.1
  []
  [biot_modulus]
    type = PorousFlowConstantBiotModulus
    solid_bulk_compliance = 1E-7
    fluid_bulk_modulus = 2E9
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '1E-13 0 0   0 1E-13 0   0 0 1E-13'
  []
[]

[Postprocessors]
  [injection_area]
    type = AreaPostprocessor
    boundary = left
    execute_on = initial
  []
  [fluid_mass0]
    type = PorousFlowFluidMass
    execute_on = timestep_begin
  []
  [fluid_mass1]
    type = PorousFlowFluidMass
    execute_on = timestep_end
  []
  [zmass_error]
    type = FunctionValuePostprocessor
    function = mass_bal_fcn
    execute_on = timestep_end
  []
  [zinj_mass]
    type = FunctionValuePostprocessor
    function = inj_mass
    execute_on = timestep_end
  []
[]

[Functions]
  [mass_bal_fcn]
    type = ParsedFunction
    value = '(c - a)'
    vars = 'a c'
    vals = 'fluid_mass1 fluid_mass0'
  []
  [inj_mass]
    type = ParsedFunction
    value = 'flux * area'
    vars = 'flux area'
    vals = 'flux  injection_area'
  []
  [flux]
    type = ParsedFunction
    vals = injection_area
    vars = area
    value = '1E-5/area'
  []
[]

[Preconditioning]
  active = mumps
  [mumps]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = ' lu       mumps'
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 10
  nl_abs_tol = 1e-9
  dt = 1
[]

[Outputs]
[]

**Comment by cpgr** [comment]
There was a similar question earlier. Did you see #17530?

**Reply by MatiasAllay** [reply]
Hi, yes I saw it thanks! But neither `PorousFlowBasicTHM` with `multiply_by_density = true` nor `PorousFlowFullySaturated` closes the mass balance. In X-Y coordinates, however, the mass balance closes.

**Comment by WilkAndy** [comment, marked as answer]
@cpgr has partially answered your question: one fundamental problem is `PorousFlowBasicTHM`.  It is not designed for this sort of thing.  I should really add this to the documentation, https://mooseframework.inl.gov/source/actions/PorousFlowBasicTHM.html , because it has tripped me up too.  Note the time derivative of BasicTHM is \dot{P}/M, while the `PorousFlowFluidMass` is calculating \phi\rho - so there's no reason why the latter should be conserved.  Simply change `PorousFlowBasicTHM` to `PorousFlowFullySaturated` (https://mooseframework.inl.gov/source/actions/PorousFlowFullySaturated.html).

@cpgr : this raises things that have been nagging me for a while.  The problem is in `PorousFlowVolumetricStrain`.  In that class we calculate div(velocity).

1. An immediate problem is that in RZ coordinates we miss the u_r/r term, ooops
2. Perhaps we should just utilize the user's calculation of total_strain, ie, replace the appropriate lines by `_vol_total_strain_qp[_qp] = _total_strain[_qp].trace();` and `total_strain_increment = _total_strain[_qp] - _total_strain_old[_qp];`.  The user has defined what they mean by strain, so why not use it?
3. The reason we calculate in the current way is so our Jacobian is exact, while if we use the user's total strain, how would we get the derivatives?  We need at least d(vol_strain)/d(grad(u)).
4. In various cases we've assumed d(vol_strain)/du = 0, but in RZ (with the u_r/r term) that's not true, boohoo, and probably the case in finite strain nightmare land too, that requires extending other classes
5. What the heck is that consistent_with_displaced_mesh that i put in?  It obviously works, but i can't figure out why.

**Reply by cpgr** [reply]
I never understood `consistent_with_displaced_mesh`!

Using the given total strain sounds attractive. I wonder if we can get the derivatives approximately correct and whether this is enough for good convergence, at least for the small strain case.

**Reply by WilkAndy** [reply]
I think one of the reasons we "custom coded" `PorousFlowVolumetricStrain` was that the mechanical coupling in PorousFlow is really only meant for small strains.  It's difficult to remember that fact when presented with an all-singing-all-dancing code like PorousFlow.   Do you think "AD" will help with the derivatives, or will it be burdensome, or not work at all?

**Reply by MatiasAllay** [reply]
Thanks Andy, I still cannot close the mass balance. But I assume that the `PorousFlowFluidMass` is not being computed somehow correctly and the injected fluid is indeed the one I am setting up in my code.

**Reply by WilkAndy** [reply]
I'm currently of the opinion that you have discovered a bug in PorousFlow, @MatiasAllay .  Some mechanically-coupled simulations will work fine, but I believe others will be buggy.    Thank you - i know it's super annoying for you, but it helps us.  I need to converse with @cpgr about the best way to fix this.

**Reply by MatiasAllay** [reply]
Hi @WilkAndy no is not annoying. Thank you and @cpgr for the great work.

**Reply by MatiasAllay** [reply]
I'd appreciate if you can come back to this post once you fix the bug :)

**Reply by WilkAndy** [reply]
Yep, will do

**Reply by WilkAndy** [reply]
Hi @MatiasAllay .  The bug fix, #18359, has been merged into "MOOSE next".   Apologies for the amount of time it took to fix - due to computer issues outside my control.  The bug fix will slowly and automatically find its way into "MOOSE master" over the next day or so.
